Add installation script for open62541 to ease installation.

eddybl opened the pull request:
This PR adds a script to install the `open62541` library in an automated fashion, including downloading the correct version.

It fixes #184 


Disclaimer: I used Copilot to help me with sections of the bash script.

eddybl committed:
Add installation script for open62541 to ease installation.

ralphlange commented on devOpcuaSup/open62541/install_open62541.sh:
```diff
@@ -0,0 +1,71 @@
+make -C "$temp_dir"/build install
+if [ $? -ne 0 ]; then
+    echo "Error: make install failed."
+    rm -rf "$temp_dir"
```
As above: Leave the `$temp_dir` in the error case for inspection?

ralphlange commented on devOpcuaSup/open62541/install_open62541.sh:
```diff
@@ -0,0 +1,71 @@
+cat <<EOL > "$SCRIPT_DIR/../../configure/CONFIG_SITE.local"
+OPEN62541 = \$(TOP)/devOpcuaSup/open62541/lib-open62541
+OPEN62541_DEPLOY_MODE = EMBED
+OPEN62541_USE_CRYPTO = YES
```
Also needs `OPEN62541_USE_XMLPARSER` here (and requires the `libxml2` devel libraries - see above), else there is no support for structured data.

eddybl replied:
Added both, but see the other comment for clarification

ralphlange commented:
Not necessarily as part of this PR, but we need to do something similar for the Windows users.

ralphlange reviewed (changes requested):
That is going into the right direction, for sure! Thanks a lot!

A few updates needed here and there.

ralphlange commented on devOpcuaSup/open62541/install_open62541.sh:
```diff
@@ -0,0 +1,71 @@
+    echo "Error: Failed to download open62541 version $OPEN62541_VERSION."
+    exit 1
+fi
+temp_dir="`mktemp -d -p "$SCRIPT_DIR"`"
```
Need to remove `$temp_dir` and the install dir to be sure to start from a clean situation.

ralphlange replied:
Ah, I didn't see `mktemp`.

I would use a fix name and remove the dir before building.
The `mktemp` is only important for parallel make etc. where two builds might be executing at the same time in the same location.

ralphlange replied:
See above. I didn't realize the `mktemp`.
I would do it in a fixed location, clean before build and leave the dir in case of error.

ralphlange commented on devOpcuaSup/open62541/install_open62541.sh:
```diff
@@ -0,0 +1,71 @@
+echo "Extracting open62541 archive..."
+mkdir -p "$SCRIPT_DIR/lib-open62541"
+tar -xzf "$SCRIPT_DIR/open62541-$OPEN62541_VERSION.tar.gz" -C "$temp_dir" --strip-components=1
+rm "$SCRIPT_DIR/open62541-$OPEN62541_VERSION.tar.gz"
```
Wouldn't it be simpler to do
`curl -L $URL | tar -xzf`

eddybl replied:
yes it would

eddybl commented on devOpcuaSup/open62541/install_open62541.sh:
```diff
@@ -0,0 +1,71 @@
+make -C "$temp_dir"/build
+if [ $? -ne 0 ]; then
+    echo "Error: make failed."
+    rm -rf "$temp_dir"
```
well, maybe instead of an random folder I should just definie a name for a specific folder to create and use...

If the script is executed, it would check if the folder already exists and delete it.

otherwise you might end up with a couple of randomly named folders if you run the script multiple times with errors

eddybl committed:
Reworked build logic.

- Use fixed build dir
- Remove build and lib dir before starting
- Remove build dir only on failure
- Combine download and extraction

eddybl committed:
Fix capitalization of cmake

eddybl commented on devOpcuaSup/open62541/README.md:
```diff
@@ -32,11 +32,24 @@ The Open62541 project is focused on the server implementation of OPC UA.
+2. Build `open62541` as a static library with security support
+3. Create the corresponding `CONFIG_SITE.local` configuration
+
+**Please note**: This requires `curl`, `Cmake` and the `openssl-devel`/`libssl-dev` library to be present.
```
I actually tried to take care of that and copied it from a couple of lines above... but I fixed both now ;)

eddybl commented on devOpcuaSup/open62541/README.md:
```diff
@@ -32,11 +32,24 @@ The Open62541 project is focused on the server implementation of OPC UA.
+2. Build `open62541` as a static library with security support
+3. Create the corresponding `CONFIG_SITE.local` configuration
+
+**Please note**: This requires `curl`, `Cmake` and the `openssl-devel`/`libssl-dev` library to be present.
```
The top of the readme states that structures are not supported for open62541 as of now and also for the "manual" build instructions I do not see any change in build configuration to allow for structure-support to be taken into account? The only thing I see is adding the line `OPEN62541_USE_XMLPARSER = YES` to the `CONFIG_SITE.local`, which I can certainly add, but not sure if that is enough? Or is this only required during runtime?

ralphlange replied:
README out of date. Thanks for pointing this out.
OPEN62541_USE_XMLPARSER = YES  -->  Structures are supported
OPEN62541_USE_XMLPARSER = NO  -->  No structures

ralphlange replied:
Background: The catalog of user-defined data descriptions is sent from the server to the client as XML.
No catalog -> no decoding.

eddybl replied:
Right, so the libxml2-dev is required for building the epics module, but not the open62541 library (which is different to libssl), and the iocs need libxml2 (without dev, once the other issue is resolved)

ralphlange replied:
That sounds right and is probably correct.

Interesting... so open62541 shifts all of that to the user, while the UA SDK does it in its plushy convenience layer, meaning that the UA SDK needs libxml2 itself, and the device support needs it for open62541.
I should have known that for a while...
I guess keeping the LIBXML2 setting in the Device Support configuration is correct for both cases. However, I should verify that the linker lines are correct for both SDKs and all linking options.

eddybl replied:
yeah so this might be relevant for #182

eddybl committed:
Add mention of `libxml2` dev files and add to local config.
